### ISSUE TITLE
caffe2 cyclical learning rate

### ISSUE THREAD

**Issue by dollysapra** [pytorch/pytorch#26310]
I recently moticed cyclical learning rate implementation and did some experiments. 
My network does not learn anything with cyclical policy (It was doing okay with other policies such as step, exp. 
What is the correct way of using it? 

I am initializing it like this  - 
optimizer.build_sgd(train_model,base_learning_rate= 0.002,
                policy="cyclical",stepsize=500, max_lr=0.05, decay=0.99,
                momentum=0.9, nesterov=1,
                weight_decay=0.0001)

I have tried with various values of stepsize, decay and learning rates. Maybe someone can help me if this is the correct way. 

Loss is also constant at 2.302585 - no change at all  through the itrations!

EDIT: I had a look at learning rate, normally i get a floating point number for learning rate through the iterations. but for cyclical policy i get "nan" for learning rate. This might be a bug. 

Thanks!

**Comment by dollysapra** [comment]
In caffe2/sgd/learning_rate_op.h
T decay =
          this->template GetSingleArgument**<int>**(arg_prefix + "decay", 1.0);

<int> shd be <float> i believe. I changed it on my machine but still getting nan learning rate..

**Comment by dollysapra** [comment]
Maybe this is my local compiler issue - 
abs(base_lr_) is always returning 0.00, I changed it to fabs(base_lr_) and learning rate is not a nan anymore. 

Lines 271-273 in caffe2/sgd/learning_rate_functors.h are now - 

```
return 1 +
        (T(fabs(max_lr_)) / T(fabs(base_lr_)) - 1) * std::max(T(0.0), (1 - x)) *
        std::pow(decay_, static_cast<int>(iter / (2 * stepsize_)));
```

Not an elegant solution - it should check for T type and then use appropriate function - or maybe there is an overloaded function available somewhere --  but it works for me now.
